---
title: "Phyloseq Preprocessing"
author: "T. Griffin"
date: "2024-03-06"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

##Load libraries
```{r load-libraries}

library(devtools)
BiocManager::install("phyloseq")

library(phyloseq)
library(tidyverse)
```

```{r figure-location}

knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "../figures/02_Preprocessing/")#send any figure output to this folder

```
#Goals

Here, we will  process the data into a phyloseq object
We will need:
1. ASV table
2. Taxonomy table
3. Track Reads (metadata)

Then, we will remove the following:

1. Remove mitochondria (BECAUSE THEY HAVE THEIR OWN 16S DNA)
2. Remove chloroplasts
3. Remove samples without "enough" readds

Finally, we will write a data file of phyloseq output.

# load data

##Load ASV Table
```{r load-asv-table}
#Load asv table

load("data/01_DADA2/DADA2_class_workflow/ASV_counts.RData")

#inspect asv_tab
head(asv_tab)[,1:5]

# Fix names

sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1)
head(samples_fixed)

#rewrite the ASV count file ot fix names
colnames(asv_tab) <- samples_fixed
str(asv_tab)

```

##Load taxonomy table
```{r load-tax-table}

tax_df <- read.table("data/01_DADA2/DADA2_class_workflow/ASV_taxonomy.tsv", sep = "\t", skip = 1)

head(tax_df)

#fix column names
colnames(tax_df) <- c("ASV", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df)


#Taxonomy table matrix
tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "ASV") %>%
  as.matrix()
```


##track reads data
```{r load-track-reads}
load("data/01_DADA2/DADA2_class_workflow/track_read_counts.RData")
#Take a look at the data
track_counts_df

head(track_counts_df)
dim(track_counts_df)

#copy metadata.csv in the terminal from /workdir/in_class_data/SalinityGradient_16S/metadata.csv. Put in the SalinityGradient_16S/data

#load in metadata
metadata_df <- read.csv("data/metadata.csv")
view(metadata_df)
dim(metadata_df)
colnames(metadata_df)

#Merge metadata_df with track_reads_df
head(metadata_df)
head(track_counts_df)

metadata_track_reads_df <- metadata_df %>%
  left_join(., track_counts_df, by = "names") %>%
  dplyr::select(-X)

#Intuition check
head(metadata_track_reads_df)
#before rewrite
row.names(metadata_track_reads_df) 

#Rewrite
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names

#check afterwards to see if it worked
row.names(metadata_track_reads_df) 

#intuition check
head(metadata_track_reads_df)
```

##Handoff to phyloseq
```{r phyloseq-handoff}
# double check that it is all good
dim(asv_tab)
dim(tax_mat)
#intuition check
stopifnot(row.names(asv_tab) == row.names(tax_mat))

#construct the phyloseq object
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE), 
                       sample_data(metadata_track_reads_df),
                                           tax_table(tax_mat))

raw_physeq

#Save this raw phyloseq object
save(raw_physeq, file = "data/02_PreProcessing/raw_physeq.RData")
```

