---
title: "Infer ASVs with DADA2"
output: html_document
date: "2024-02-21"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "../figures/01_DADA2/")#send any figure output to this folder
```


# Goals of this file

1. use raw fastq files and generate the quality plots to assess quality of reads
2. filter and trim out bad sequences and bases from our raw sequencing files
3. Write out fastq files with high quality sequences
4. Evaluate the quality from our filter and trim

#load libraries
```{r Load-libraries}

library(devtools)

library(dada2)
```

# Load Data
```{r load-data}
# Set the raw fastq paths to the raw sequencing files
#Path to the fastq files
raw_fastqs_path <- "data/01_DADA2/01_rawgzipped_fastqs/"
raw_fastqs_path

#What files are in this path? Intuition Check
list.files(raw_fastqs_path)

#how many files are there?
str(list.files(raw_fastqs_path))

#Create vector of forward reads
#forward_reads <- 
  
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)
head(forward_reads)

#Create vector of reverse reads
reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
head(reverse_reads)

```

# Quality Plots
```{r raw-quality-plots}
# randomly select two samples for each dataset to evaluate
random_samples <- sample(1:length(reverse_reads), size = 2)
random_samples

#calculate and plot quality of htese two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])
```

Prepare a placeholder for filtered reads
```{r prep-filtered-sequences}
#vector for samples, extract sample name from files
samples <- sapply(strsplit(basename(forward_reads), "_"), `[`,1)
head(samples)

# Place filtered reads into filtered_fastqs_path
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

# Create 2 variables: filtered_F, filtered_R
filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R1_filtered.fastq.gz"))
length(filtered_forward_reads)

filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R2_filtered.fastq.gz"))
head(filtered_reverse_reads)
```



#Filter and Trim
```{r filter-and-trim}
?filterAndTrim



```

