---
title: "Phyloseq Preprocessing"
author: "T. Griffin"
date: "2024-03-06"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## R Markdown

##Load libraries
```{r load-libraries}

library(devtools)
#BiocManager::install("phyloseq")

library(phyloseq)
library(tidyverse)
```

```{r figure-location}

knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "../figures/02_Preprocessing/")#send any figure output to this folder

```
#Goals

Here, we will  process the data into a phyloseq object
We will need:
1. ASV table
2. Taxonomy table
3. Track Reads (metadata)

Then, we will remove the following:

1. Remove mitochondria (BECAUSE THEY HAVE THEIR OWN 16S DNA)
2. Remove chloroplasts
3. Remove samples without "enough" readds

Finally, we will write a data file of phyloseq output.

# load data

##Load ASV Table
```{r load-asv-table}
#Load asv table

load("/local/workdir/tag92/git_repos/SalinityGradient_16S/data/01_DADA2/ASV_counts.RData")

#inspect asv_tab
head(asv_tab)[,1:5]

# Fix names

sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1)
head(samples_fixed)

#rewrite the ASV count file ot fix names
colnames(asv_tab) <- samples_fixed
str(asv_tab)

```

##Load taxonomy table
```{r load-tax-table}

tax_df <- read.table("/local/workdir/tag92/git_repos/SalinityGradient_16S/data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", skip = 1)

head(tax_df)

#fix column names
colnames(tax_df) <- c("ASV", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df)


#Taxonomy table matrix
tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "ASV") %>%
  as.matrix()
```


##track reads data
```{r load-track-reads}
load("/local/workdir/tag92/git_repos/SalinityGradient_16S/data/01_DADA2/track_read_counts.RData")
#Take a look at the data
track_counts_df

head(track_counts_df)
dim(track_counts_df)

#copy metadata.csv in the terminal from /workdir/in_class_data/SalinityGradient_16S/metadata.csv. Put in the SalinityGradient_16S/data

#load in metadata
metadata_df <- read.csv("/local/workdir/tag92/git_repos/SalinityGradient_16S/data/metadata.csv")
view(metadata_df)
dim(metadata_df)
colnames(metadata_df)

#Merge metadata_df with track_reads_df
head(metadata_df)
head(track_counts_df)

metadata_track_reads_df <- metadata_df %>%
  left_join(., track_counts_df, by = "names") %>%
  dplyr::select(-X)

#Intuition check
head(metadata_track_reads_df)
#before rewrite
row.names(metadata_track_reads_df) 

#Rewrite
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names

#check afterwards to see if it worked
row.names(metadata_track_reads_df) 

#intuition check
head(metadata_track_reads_df)
```

##Handoff to phyloseq
```{r phyloseq-handoff}
# double check that it is all good
dim(asv_tab)
dim(tax_mat)
#intuition check
stopifnot(row.names(asv_tab) == row.names(tax_mat))

#construct the phyloseq object
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE), 
                       sample_data(metadata_track_reads_df),
                                           tax_table(tax_mat))

raw_physeq

#Save this raw phyloseq object
save(raw_physeq, file = "/local/workdir/tag92/git_repos/SalinityGradient_16S/data/02_PreProcessing/raw_physeq.RData")

#Symbolic links are for large files; however, you can actually cp files that are within the size threshold
```

#Clean up the data

Remove: mitochondria
chloroplasts
```{r rm-mito-and-chloro}
#make new physeq without mito. and chloro.

#remind myself of tax table; check table for both chloro and mitocho
#remove chloro

noChloros_physeq <- raw_physeq %>%
  #removing the chloro
  subset_taxa(Order != "Chloroplast" | is.na(Order))
  
#How many taxa were chloroplasts?
num_chloros_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloros_ASVs

#Intuition check
#noChloros_physeq %>%
 # tax_table() %>%
  #data.frame() %>%
  #view()

#remove mitochondria
nochlorosmitos_physeq <- noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

#How many mitochondrial ASVs?
num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(nochlorosmitos_physeq)
num_mito_ASVs

#How many total asvs were removed>
ntaxa(raw_physeq) - ntaxa(nochlorosmitos_physeq)

#what proportion of ASVs were kept?
ntaxa(nochlorosmitos_physeq)/ntaxa(raw_physeq)



```


# Evaluate and remove any of the control samples
# take a look at the negative controls and then make a decision about whether or not to remove the ASVs that we found in our controls

1. Negative controls
2. ASVs found within the negative controls and their distribution in the samples
3. Evaluate the mock community

##Negative controls
```{r neg-controls}
#create a vector of samples that were negative controls
control_samples <- c("WaterControl", "022um-Control", "3um-Control", "DNA-Ext_Control")

#make a new phyloseq of only control samples
control_physeq <- 
  nochlorosmitos_physeq %>%
  subset_samples(., names %in% control_samples) %>%
  # force remove ASVs with 0 counts 
  prune_taxa(taxa_sums(.) > 0, .)

  
control_ASvs <-
control_physeq %>%
  tax_table() %>%
  data.frame() %>%
dplyr::select(ASV) %>%
  as.vector()

#visualize raw abundance values of control ASVs
control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")

######### Check



#We decided that we would keep ASV_18 and ASV_56, everything else is trash
controlASVs_toREMOVE <- dplyr::setdiff(control_ASvs$ASV, c("ASV_18", "ASV_56"))
#double check
length(controlASVs_toREMOVE)

##Make my new phyloseq object
noControlsChlorosMitos_physeq <-
  nochlorosmitos_physeq %>%
  #Actually remove the control ASVs (except 18, 56)
  subset_taxa(., !(ASV %in% controlASVs_toREMOVE)) %>%

#Remove control samples
subset_samples(., !(names %in% control_samples))

#inspect no controls object

#after filtering how many ASVs were removed?
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noControlsChlorosMitos_physeq)
```

# Evaluate accuracy in mock community
```{r}
# Make mock phyloseq object 
mock_physeq <- 
  noControlsChlorosMitos_physeq %>%
  subset_samples(., names == "MockZymoPos") %>%
  prune_taxa(taxa_sums(.) > 0, .)

# Inspect it 
mock_physeq

tax_table(mock_physeq)

# write a messages summarizing this
cat("DADA2 inferred", ntaxa(mock_physeq), "ASVs present in the mock community.")

####### Load in the mock reference sequence from Zymo 
mock_ref <- getSequences("/local/workdir/tag92/git_repos/SalinityGradient_16S/data/02_PreProcessing/mock_amplicons.fasta")
names(mock_ref)

### Pull out seqs from Sequenced Mock community
mock_seqs <- 
  mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq) 
# Inpsect
head(mock_seqs)

matched_asvs <- 
  mock_seqs %>%
  rowwise() %>%
  # check each ASV if they match, TRUE, if not = FALSE
  mutate(Match = any(grepl(ASVseq, mock_ref))) %>%
  # Create a vector of names that matched 
  pull(ASV)

cat(length(matched_asvs), "ASVs were exact matches to the expected reference sequence.")

# How many sequences were in the mock? 
mock_SeqDepth <- 
  mock_physeq %>%
  otu_table() %>%
  data.frame() %>%
  colSums()


# Who is in the sequenced mock community? 
mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(Genus, Species)

#What are the mock counts in the actual samples 
noControlsChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% matched_asvs) %>%
  otu_table() %>%
  t()
```

## Remove mock community
```{r}
# Make a new phyloseq object without mock and its asvs
noControlsMock_ChlorosMitos_physeq <- 
  noControlsChlorosMitos_physeq %>%
  # remove the mock community 
  subset_samples(., names != "MockZymoPos") %>%
  # remove the ASVs from the mock 
  subset_taxa(., !(ASV %in% matched_asvs))

noControlsMock_ChlorosMitos_physeq
```

# Evaluate Sequencing Depth
```{r}
noControlsMock_ChlorosMitos_physeq

seqSums_df <- 
  noControlsMock_ChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_track_reads_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)

head(seqSums_df)
#show the depth of samples 
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = station)) + 
  geom_bar(stat = "identity") 

#Density plot 
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = station)) +
  geom_density(alpha = 0.5)

```

# remove samples with few reads
```{r}
min(sample_sums(noControlsMock_ChlorosMitos_physeq))

#Remove 20210615-MA-SCB2F
raw_preprocessed_physeq <- 
  noControlsMock_ChlorosMitos_physeq %>%
  subset_samples(., names != "20210615-MA-SCB2F")

#new min seq depth
min(sample_sums(raw_preprocessed_physeq))
```

# Save preprocessed phyloseq object
```{r}
save(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")
```

# Session information
```{r}
devtools::session_info()
```

